gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/ApiReactorHandlerFactory.java

```diff
@@ -274,20 +273,15 @@ protected io.gravitee.gateway.reactor.handler.context.ExecutionContextFactory v3
         return executionContextFactory;
     }
 
-    private ExecutionContextFactory executionContextFactory(
-        Api api,
-        ComponentProvider componentProvider,
-        DefaultReferenceRegister referenceRegister
-    ) {
-        final ExecutionContextFactory executionContextFactory = new ExecutionContextFactory(componentProvider);
-        executionContextFactory.addTemplateVariableProvider(new ApiTemplateVariableProvider(api));
-        executionContextFactory.addTemplateVariableProvider(referenceRegister);
-        applicationContext
-            .getBean(ApiTemplateVariableProviderFactory.class)
-            .getTemplateVariableProviders()
-            .forEach(executionContextFactory::addTemplateVariableProvider);
+    private List<TemplateVariableProvider> templateVariableProviders(Api api, DefaultReferenceRegister referenceRegister) {
+        List<TemplateVariableProvider> templateVariableProviders = new ArrayList<>();
+        templateVariableProviders.add(new ApiTemplateVariableProvider(api));
+        templateVariableProviders.add(referenceRegister);
+        templateVariableProviders.addAll(
+            applicationContext.getBean(ApiTemplateVariableProviderFactory.class).getTemplateVariableProviders()
```

Comment:
I know that it is a copy of existing code, but is it not too expensive to look after beans with Spring ?
This will be executed for each of the incoming request.

Also, it seems that ApiTemplateVariableProviderFactory.getTemplateVariableProviders() is not synchronized so the providers may be initialized multiple times.

Reply:
I think this initialization stuff is made when the api is deployed not per request so it should not be an issue.